src/nm-checkpoint.c

```diff
@@ -206,15 +206,93 @@ find_settings_connection (NMCheckpoint *self,
 	return sett_conn;
 }
 
+static gboolean
+restore_and_activate_connection (NMCheckpoint *self,
+                                 DeviceCheckpoint *dev_checkpoint)
+{
+	NMCheckpointPrivate *priv = NM_CHECKPOINT_GET_PRIVATE (self);
+	NMSettingsConnection *connection;
+	gs_unref_object NMAuthSubject *subject = NULL;
+	GError *local_error = NULL;
+	gboolean need_update, need_activation;
+
+	connection = find_settings_connection (self,
+	                                       dev_checkpoint,
+	                                       &need_update,
+	                                       &need_activation);
+	if (connection) {
+		if (need_update) {
+			_LOGD ("rollback: updating connection %s",
+			       nm_settings_connection_get_uuid (connection));
+			nm_settings_connection_update (connection,
+			                               dev_checkpoint->settings_connection,
+			                               NM_SETTINGS_CONNECTION_PERSIST_MODE_DISK,
+			                               NM_SETTINGS_CONNECTION_COMMIT_REASON_NONE,
+			                               "checkpoint-rollback",
+			                               NULL);
+		}
+	} else {
+		/* The connection was deleted, recreate it */
+		_LOGD ("rollback: adding connection %s again",
+		       nm_connection_get_uuid (dev_checkpoint->settings_connection));
+
+		connection = nm_settings_add_connection (nm_settings_get (),
+		                                         dev_checkpoint->settings_connection,
+		                                         TRUE,
+		                                         &local_error);
+		if (!connection) {
+			_LOGD ("rollback: connection add failure: %s", local_error->message);
+			g_clear_error (&local_error);
+			return FALSE;
+		}
+		need_activation = TRUE;
+	}
+
+	if (need_activation) {
+		_LOGD ("rollback: reactivating connection %s",
+		       nm_settings_connection_get_uuid (connection));
+		subject = nm_auth_subject_new_internal ();
+
+		/* Disconnect the device if needed. This necessary because now
+		 * the manager prevents the reactivation of the same connection by
+		 * an internal subject. */
+		if (   nm_device_get_state (dev_checkpoint->device) > NM_DEVICE_STATE_DISCONNECTED
+		    && nm_device_get_state (dev_checkpoint->device) < NM_DEVICE_STATE_DEACTIVATING) {
+			nm_device_state_changed (dev_checkpoint->device,
+						 NM_DEVICE_STATE_DEACTIVATING,
+						 NM_DEVICE_STATE_REASON_NEW_ACTIVATION);
```

Comment:
Wrong indentation. Also, this is complained by checkpatch.pl. Are you running that as commit-hook? That would be a good idea.

Reply:
thanks, fixed